TulparAudio/internal/source/SourceCollection.cpp

```diff
@@ -0,0 +1,1097 @@
+
+    audio::Buffer buffer = GetSourceActiveBuffer(source);
+
+    ALint sampleOffset = 0;
+
+    {
+        std::vector<audio::Buffer> buffers = GetSourceActiveBuffers(source);
+        uint64_t offsetNs = offset.count();
+        uint32_t i = 0;
+
+        while (offsetNs > 0)
+        {
+            // get buffer
+            audio::Buffer const& buffer = buffers.at(i);
+
+            // check how much time from this buffer we can consume
+            uint64_t const consumedTime = std::min(
+                offsetNs
+                , static_cast<uint64_t>(buffer.GetDuration().count())
+            );
+
+            // calculate time based on this buffer properties
+            uint32_t const sampleCount = buffer.GetFrequencyHz()
+                * (static_cast<double>(consumedTime) * 1e-9);
+
+            sampleOffset += std::min(sampleCount, buffer.GetSampleCount());
+
+            // removed consumed time
+            offsetNs -= consumedTime;
+            ++i;
+        }
+    }
+
+
+    // clear error state
+    ALenum alErr = alGetError();
+
+    alSourcei(static_cast<ALuint>(source), AL_SAMPLE_OFFSET, sampleOffset);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        LOG_AUDIO->Warning("Source #{}: set playback position: {:#x}", source, alErr);
+    }
+
+    return AL_NO_ERROR == alErr;
+}
+
+float SourceCollection::GetSourcePlaybackProgress(SourceHandle source) const
+{
+    float result = 0.0f;
+
+    // clear error state
+    ALenum alErr = alGetError();
+
+    ALint sampleOffset;
+    alGetSourcei(static_cast<ALuint>(source), AL_SAMPLE_OFFSET, &sampleOffset);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR == alErr)
+    {
+        audio::Source::State const state = GetSourceState(source);
+
+        if (sampleOffset > 0
+            && ((audio::Source::State::Playing == state)
+                || (audio::Source::State::Paused == state))
+        )
+        {
+            result = static_cast<float>(sampleOffset) / static_cast<float>(m_sourceMeta.at(source).activeSampleCount);
+        }
+    }
+    else
+    {
+        LOG_AUDIO->Warning("Source #{}: get playback progress: {:#x}", source, alErr);
+    }
+
+    return result;
+}
+
+bool SourceCollection::SetSourcePlaybackProgress(SourceHandle source, float value)
+{
+    LOG_AUDIO->Debug("Source #{}: set playback progress {}%", source, value);
+
+    audio::Buffer buffer = GetSourceActiveBuffer(source);
+
+    ALint sampleOffset = static_cast<ALint>(std::round(static_cast<float>(m_sourceMeta[source].activeSampleCount) * value));
+
+    // clear error state
+    ALenum alErr = alGetError();
+
+    alSourcei(static_cast<ALuint>(source), AL_SAMPLE_OFFSET, sampleOffset);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        LOG_AUDIO->Warning("Source #{}: set playback progress: {:#x}", source, alErr);
+    }
+
+    return AL_NO_ERROR == alErr;
+}
+
+audio::Source::State SourceCollection::GetSourceState(SourceHandle source) const
```

Comment:
Shouldn't it be static?

Reply:
Answered in https://github.com/Godlike/Tulpar/pull/6#discussion_r244268061